### ISSUE TITLE
ESLint compatibility: native rules report false positives for common JS/TS patterns

### ISSUE THREAD

**Issue by fireairforce** [utooland/utoo-lint#1191]
## Summary

When `@utoo/lint` is used as an ESLint-compatible replacement, several native rules report diagnostics for common JavaScript and TypeScript patterns that the equivalent ESLint rules accept.

These differences can create a large number of migration-only errors even when the same files pass ESLint with the same rule severities.

## Environment

- `@utoo/lint`: 0.3.4
- ESLint: 8.35.0
- `@typescript-eslint/eslint-plugin`: 5.62.x
- Node.js: 22.x

### 1. Recursive closure reported by `no-use-before-define`

```js
const visit = node => (node ? visit(node.next) : null);
visit(root);
```

With `no-use-before-define: "error"`:

- ESLint: no diagnostic
- utoo-lint: `'visit' was used before it was defined.`

The reference is inside a function that runs after the variable has been initialized. The same issue affects methods or callbacks nested inside a variable initializer, for example a store object that references the exported store from its action methods.

The native implementation appears to treat any reference whose source span is inside the initializer as an initializer-time reference, without excluding references across a nested function scope.

### 2. `@typescript-eslint/no-unused-vars` reports non-variable declarations

```ts
export enum Mode {
  Read = 'read',
  Write = 'write',
}

export type EnvMap = {
  [key in 'DEV' | 'PROD']: number;
};

export const Component = wrap(function DisplayName() {
  return null;
});
```

With `@typescript-eslint/no-unused-vars: "error"`:

- ESLint: no diagnostics
- utoo-lint reports the enum members, mapped-type key, and named function expression as unused variables

Unused catch bindings are also reported by default:

```ts
try {
  work();
} catch (error) {
  recover();
}
```

ESLint's default for `caughtErrors` is `"none"`, while utoo-lint currently behaves as if it were `"all"` when the rule is enabled with severity only.

### 3. Async array callbacks reported by `array-callback-return`

```js
await Promise.all(
  items.map(async item => {
    await consume(item);
  }),
);
```

With `array-callback-return: "error"`:

- ESLint: no diagnostic
- utoo-lint: `Expected to return a value in array callback.`

ESLint excludes async and generator callbacks from this check. The native implementation currently checks the callback body without first excluding those function kinds.

### 4. ESLint suppression comments do not suppress native diagnostics

```js
function normalize(value) {
  value = String(value); // eslint-disable-line no-param-reassign
  return value;
}
```

With `no-param-reassign: "error"`:

- ESLint: diagnostic is suppressed
- utoo-lint: diagnostic remains in the result

The JavaScript compatibility layer applies ESLint directives to custom-rule messages, but native diagnostics appear to bypass the same filtering path.

## Expected behavior

For supported ESLint rule IDs, severity-only configuration should preserve ESLint's documented defaults and avoid additional diagnostics for syntax patterns accepted by the corresponding ESLint rule. Existing `eslint-disable`, `eslint-disable-line`, and `eslint-disable-next-line` comments should also suppress native diagnostics.

## Suggested regression coverage

- Distinguish direct initializer self-references from references inside nested function scopes.
- Match `@typescript-eslint/no-unused-vars` defaults, especially `caughtErrors: "none"`.
- Exclude enum members, mapped-type keys, and named function/class expression names from unused-variable diagnostics.
- Skip async and generator callbacks in `array-callback-return`, matching ESLint.
- Apply ESLint suppression directives to native diagnostics returned through `Linter`, `ESLint`, and `CLIEngine` compatibility APIs.
- Add differential tests that run these fixtures through both ESLint and utoo-lint and compare rule IDs.
